### PR TITLE
OpenFlow rule filter

### PR DESCRIPTION
**Pull request by dvandra**
OpenFlow rule filter: By Gremlin request with "OR" condition, using "as/select" method. (without removing "regex" and not using "HasEither".)

### CODE REVIEW

**Comment by lebauce** [comment]
Please go easy on closing/reopening PRs :-)

File: statics/js/components/rule-detail.js [hunk 1]
```diff
@@ -293,29 +293,30 @@ var BridgeLayout = (function () {
   BridgeLayout.prototype.extract = function () {
     var itfs = {};
     var rules = [];
-    var rulesUUID = new Set();
+    var rulesUUID = new Set();          // added on check tests, by p.c.
     var groupsUUID = new Set();
-    var groups = [];
+    var groups = [];
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: Indentation in the whole file need to be fixed

File: statics/js/components/rule-detail.js [hunk 2]
```diff
@@ -580,78 +581,130 @@ Vue.component('groups-detail', {
...
+  },
+
   props: {
     bridge: {
       type: Object,
       required: true
-    },
-    graph: {
+      },
+    realgraph: {
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: I would prefer to use graph here (I know that's me who used real for the draft :)

File: statics/js/components/rule-detail.js [hunk 3]
```diff
@@ -580,78 +581,130 @@ Vue.component('groups-detail', {
...
     }
   },
 
   data: function() {
-    return { memoBridgeLayout:null };
+    return {
+      value: "",
+      memoBridgeLayout:null,
+      filters: {},
+      graph: {
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: and subgraph here as it comes from a gremlin request

File: statics/js/components/rule-detail.js [hunk 4]
```diff
@@ -666,45 +719,128 @@ Vue.component('rule-detail', {
...
+      if (has.length > 0) {
+	        for (var k in this.filters) {
+                //queryRules += ".Has('filters', regex('.*" + k + "=" + this.filters[k] + ".*')).As('"k"')";
+                query += queryRules + ".Has('filters', regex('.*" + k + "=" + this.filters[k] + ".*')).As('" + k + "').";
+	        }
+      }
+      else{
+	    query += queryRules + ".As('ofrules')."
+      }
+      //queryRules += ".As('ofrules')"
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: comment needs to be removed

File: statics/js/components/rule-detail.js [hunk 5]
```diff
@@ -666,45 +718,115 @@ Vue.component('rule-detail', {
...
+            }    
+        };
+        var handleUpdate = function(n) {
+            if ((n.metadata.Type == 'ofgroup' && self.memoBridgeLayout.groupsUUID.has(n.metadata.UUID)) ||
+                (n.metadata.Type == 'ofrule' && self.memoBridgeLayout.rulesUUID.has(n.metadata.UUID))) {
+                self.memoBridgeLayout = null;
+            }
+        };
+        this.handler = {
+        onEdgeAdded: handle,
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: indentation seems to be still not correct

File: statics/js/components/rule-detail.js [hunk 6]
```diff
@@ -666,45 +718,115 @@ Vue.component('rule-detail', {
...
+        onEdgeDeleted: handle,
+        onNodeUpdated:handleUpdate
+        };
+        this.realgraph.addHandler(this.handler);
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: You changed according to my comment thx, but I'm not sure this is still working ? I think more modification are required

File: statics/js/components/rule-detail.js [hunk 7]
```diff
@@ -666,45 +718,115 @@ Vue.component('rule-detail', {
...
+      var i = 0;
+      for (var k in this.filters) {
+        has += "'filters', regex('.*" + k + "=" + this.filters[k] + ".*')";
+	    list[i] = k;
+	    i +=1;
+      }
+      var query = queryBridge + "." + queryPorts + "." ;
+      if (has.length > 0) {
+	        for (var k in this.filters) {
+            query += queryRules + ".Has('filters', regex('.*" + k + "=" + this.filters[k] + ".*')).As('" + k + "').";
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: indent

File: statics/js/components/rule-detail.js [hunk 8]
```diff
@@ -666,45 +718,115 @@ Vue.component('rule-detail', {
...
+      var queryBridge = "G.V('" + self.bridge.id + "').As('bridge')";
+      var queryPorts = queryBridge + ".Out().Has('Type', 'ovsport').As('ovsports')";
+      var queryRules = queryBridge + ".Out().Has('Type', 'ofrule')";
+      var has = "";
+      var list = [];
+      var i = 0;
+      for (var k in this.filters) {
+        has += "'filters', regex('.*" + k + "=" + this.filters[k] + ".*')";
+	    list[i] = k;
+	    i +=1;
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: indent

File: statics/js/components/rule-detail.js [hunk 9]
```diff
@@ -666,45 +718,115 @@ Vue.component('rule-detail', {
...
+	        for (var k in this.filters) {
+            query += queryRules + ".Has('filters', regex('.*" + k + "=" + this.filters[k] + ".*')).As('" + k + "').";
+	        }
+      }
+      else{
+	    query += queryRules + ".As('ofrules')."
+      }
+      query += "Select('bridge', 'ovsports'";
+      if (has.length > 0){
+      	for (var p in this.filters){
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: indent, please use an auto indent tool

File: statics/js/components/rule-detail.js [hunk 10]
```diff
@@ -580,78 +580,130 @@ Vue.component('groups-detail', {
...
     }
   },
 
   data: function() {
-    return { memoBridgeLayout:null };
+    return {
+      value: "",
+      memoBridgeLayout:null,
+      filters: {},
+      Subgraph: {
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: s/Subgraph/subgraph/ and you need to adapt to whole code accordingly

File: statics/js/components/rule-detail.js (hunk 5 above)
Comment (dvandra) [reply statics/js/components/rule-detail.js]: Got it!

File: statics/js/components/rule-detail.js [hunk 11]
```diff
@@ -74,27 +74,27 @@ function computeAction(rule, element) {
...
-        break;
-      case 'local':
-        summary.port = LOCAL_PORT;
-        break;
-      case 'in_port':
-        summary.port = rule.inPort;
-        break;
-    }
-    rule.outAction.push(summary);
+	var summary = { action: action };
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: You need to keep space indentation. Here you changed to tabulation indentation

**Comment by nplanel** [comment]
run tests

**Comment by nplanel** [comment]
run tests

**Comment by nplanel** [comment]
run tests

File: statics/js/components/rule-detail.js [hunk 12]
```diff
@@ -580,65 +579,80 @@ Vue.component('groups-detail', {
...
 
+
 /** Vue component showing the rules associated to a bridge */
 Vue.component('rule-detail', {
+
+  mixins: [apiMixin],
+
   template: '\
-<div class="rules-detail flow-ops-panel" v-if="Object.keys(layout.structured).length > 0">\
+    <div class="rules-detail flow-ops-panel" v-if="Object.keys(layout.structured).length > 0">\
```
Comment (safchain) [inline statics/js/components/rule-detail.js]: i don't know what changed but the background color is not the same as previously. Can you use the same "style" as flow  filtering for the filtering button/footer.

**Comment by safchain** [comment]
Applying the following diff will fix the css issue
https://paste.fedoraproject.org/paste/l6t-eU9jmitTZFiVtde71Q

**Comment by pierrecregut** [comment]
Nice, but it breaks some functionality of the basic component: the subgraph does not have the interfaces linked to the ovsports. The result is that the port names are no more displayed. I will open a bug.

**Pull request merged**